Introduce ActionMap to Terminal Settings Model

carlos-zamora opened the pull request:
## Summary of the Pull Request
This entirely removes `KeyMapping` from the settings model, and builds on the work done in #9543 to consolidate all actions (key bindings and commands) into a unified data structure (`ActionMap`).

## References
#9428 - Spec
#6900 - Actions page

Closes #7441

## Detailed Description of the Pull Request / Additional comments
The important thing here is to remember that we're shifting our philosophy of how to interact/represent actions. Prior to this, the actions arrays in the JSON would be deserialized twice: once for key bindings, and again for commands. By thinking of every entry in the relevant JSON as a `Command`, we can remove a lot of the context switching between working with a key binding vs a command palette item.

#9543 allows us to make that shift. Given the work in that PR, we can now deserialize all of the relevant information from each JSON action item. This allows us to simplify `ActionMap::FromJson` to simply iterate over each JSON action item, deserialize it, and add it to our `ActionMap`.

Internally, our `ActionMap` operates as discussed in #9428 by maintaining a `_KeyMap` that points to an action ID, and using that action ID to retrieve the `Command` from the `_ActionMap`. Adding actions to the `ActionMap` automatically accounts for name/key-chord collisions. A `NameMap` can be constructed when requested; this is for the Command Palette.

Querying the `ActionMap` is fairly straightforward. Helper functions were needed to be able to distinguish an explicit unbinding vs the command not being found in the current layer. Internally, we store explicitly unbound names/key-chords as `ShortcutAction::Invalid` commands. However, we return `nullptr` when a query points to an unbound command. This is done to hide this complexity away from any caller.

The command palette still needs special handling for nested and iterable commands. Thankfully, the expansion of iterable commands is performed on an `IMapView`, so we can just expose `NameMap` as a consolidation of `ActionMap`'s `NameMap` with its parents. The same can be said for exposing key chords in nested commands.

## Validation Steps Performed

All local tests pass.

carlos-zamora commented on src/cascadia/TerminalSettingsModel/KeyChordSerialization.cpp:
```diff
@@ -301,9 +301,12 @@ winrt::hstring KeyChordSerialization::ToString(const KeyChord& chord)
 
 KeyChord ConversionTrait<KeyChord>::FromJson(const Json::Value& json)
 {
-    const std::string keyChordText{ json.asString() };
```
I should move all of this to #9543

carlos-zamora commented on src/cascadia/TerminalSettingsModel/Command.cpp:
```diff
@@ -211,7 +218,23 @@ namespace winrt::Microsoft::Terminal::Settings::Model::implementation
+                return result;
+            }
+
+            // GH#4239 - If the user provided more than one key
+            // chord to a "keys" array, warn the user here.
+            // TODO: GH#1334 - remove this check.
+            const auto keysJson{ json[JsonKey(KeysKey)] };
+            if (keysJson.isArray() && keysJson.size() > 1)
+            {
+                warnings.push_back(SettingsLoadWarnings::TooManyKeysForChord);
+            }
+            else
+            {
+                JsonUtils::GetValueForKey(json, KeysKey, result->_Keys);
```
This should be moved to #9543

carlos-zamora replied:
No need

carlos-zamora commented on src/cascadia/TerminalSettingsModel/CascadiaSettingsSerialization.cpp:
```diff
@@ -248,57 +248,6 @@ winrt::Microsoft::Terminal::Settings::Model::CascadiaSettings CascadiaSettings::
         // If this throws, the app will catch it and use the default settings
         resultPtr->_ValidateSettings();
 
-        // GH 3855 - Gathering Data on custom profiles to inform better defaults
```
Covered this with Dustin. This was a quick-n-easy way to gather data on the key bindings. Since we removed `KeyMapping`, we don't actually have a way to serialize our key bindings (yet).

Makes more sense to just outright remove this, then we leverage the settings model to gather data better later (if desired).

carlos-zamora commented on src/cascadia/TerminalSettingsModel/ActionMapSerialization.cpp:
```diff
@@ -0,0 +1,78 @@
+                continue;
+            }
+
+            // TODO CARLOS: We get a serialization error here, but only on settings reload.
+            //              "The application called an interface that was marshalled for a different thread."
+            AddAction(*Command::FromJson(cmdJson, warnings));
```
Yeah, I don't understand what's causing this. Anybody got any ideas?

carlos-zamora replied:
It was the property changed events on Command. Removed.

carlos-zamora reviewed (commented):
As described in the PR Description, I can't run the local tests for the settings model, for some reason. Here's the error I get:
```
3>Microsoft.Terminal.Settings.Model.Lib.lib(init.obj) : error LNK2005: DllMain already defined in MSVCRTD.lib(dll_dllmain_stub.obj)
```

Anybody got any ideas?

zadjii-msft commented on src/cascadia/TerminalSettingsModel/ActionAndArgs.cpp:
```diff
@@ -200,7 +200,7 @@ namespace winrt::Microsoft::Terminal::Settings::Model::implementation
     // Return Value:
     // - a deserialized ActionAndArgs corresponding to the values in json, or
-    //   null if we failed to deserialize an action.
+    //   an "invalid" action if we failed to deserialize an action.
```
note to self: this was used in two places:
```
src\cascadia\TerminalSettingsModel\Command.cpp:
  187:                 auto actionAndArgs = ActionAndArgs::FromJson(actionJson, warnings);

src\cascadia\TerminalSettingsModel\KeyMappingSerialization.cpp:
  134:             auto actionAndArgs = ActionAndArgs::FromJson(commandVal, warnings);

```
make sure those both do the right thing with Invalid instead of null now

carlos-zamora replied:
Now there's only 1 spot (`Command`)! It was fine before, but a little redundant. Pushing a minor change to make it less redundant and leverage that `FromJson` can return an invalid action instead of nullptr.

zadjii-msft commented on src/cascadia/TerminalApp/TerminalPage.cpp:
```diff
@@ -919,23 +913,23 @@ namespace winrt::TerminalApp::implementation
-        const auto actionAndArgs = _settings.KeyMap().TryLookup(kc);
-        if (actionAndArgs && (actionAndArgs.Action() == ShortcutAction::CloseTab || actionAndArgs.Action() == ShortcutAction::NextTab || actionAndArgs.Action() == ShortcutAction::PrevTab || actionAndArgs.Action() == ShortcutAction::ClosePane))
+        const auto cmd{ _settings.ActionMap().GetActionByKeyChord(kc) };
+        if (cmd && (cmd.Action().Action() == ShortcutAction::CloseTab || cmd.Action().Action() == ShortcutAction::NextTab || cmd.Action().Action() == ShortcutAction::PrevTab || cmd.Action().Action() == ShortcutAction::ClosePane))
```
we gotta do something about `Action().Action()`, because that's mental to me. Plus, given how frequently we interchange `cmd`, `args`, `action`, it's probably best to make sure each of the layers has a unique property name, so we immediately know what `foo.Action()` is

carlos-zamora replied:
Changed to `command.ActionAndArgs().Action()`

zadjii-msft commented on src/cascadia/TerminalSettingsModel/GlobalAppSettings.cpp:
```diff
@@ -326,24 +312,15 @@ void GlobalAppSettings::LayerJson(const Json::Value& json)
             // warnings generated from parsing these keybindings, add them to our
             // list of warnings.
             _keybindingsWarnings.insert(_keybindingsWarnings.end(), warnings.begin(), warnings.end());
-
```
very happy we don't have to do this dance anymore

zadjii-msft commented on src/cascadia/TerminalSettingsModel/KeyChordSerialization.cpp:
```diff
@@ -103,10 +104,8 @@ static const std::unordered_map<std::wstring_view, int32_t> vkeyNamePairs {
 // Return Value:
 // - a newly constructed KeyChord
-KeyChord KeyChordSerialization::FromString(const winrt::hstring& hstr)
+static KeyChord _FromString(const std::wstring& wstr)
```
I think if you make this a std::wstring_view, then you can pass the hstring into it directly, without constructing an intermediate wstring

zadjii-msft commented on src/cascadia/TerminalSettingsModel/KeyChordSerialization.cpp:
```diff
@@ -282,5 +286,45 @@ winrt::hstring KeyChordSerialization::ToString(const KeyChord& chord)
+    {
+        const std::string keyChordText{ json.isString() ?
+                                            json.asString() :
+                                            json[0].asString() };
```
danger: what if someone does 
```json
"keys": { "foo": true, "bar": "you get the picture" }
```

I'd expect that doesn't throw `winrt::hresult_invalid_argument`

zadjii-msft commented on src/cascadia/TerminalSettingsModel/KeyChordSerialization.cpp:
```diff
@@ -282,5 +286,45 @@ winrt::hstring KeyChordSerialization::ToString(const KeyChord& chord)
+
+bool ConversionTrait<KeyChord>::CanConvert(const Json::Value& json)
+{
+    return json.isString() || (json.isArray() && json.size() == 1);
```
I suppose it'll be saved by this, but _still_

zadjii-msft commented on src/cascadia/TerminalSettingsModel/ActionMap.idl:
```diff
@@ -0,0 +1,30 @@
+
+namespace Microsoft.Terminal.Settings.Model
+{
+    [default_interface] runtimeclass ActionMap {
```
Tons of dead code up in here

zadjii-msft reviewed (commented):
Alright, so I'm at 42/52. Probably 30 of those were fully just renaming variables, but I am making progress on this. I'll keep at it.

Maybe for the `Action().Action()` I've got three equally bad ideas:
* `command.Action().Type()`
* `command.ActionAndArgs().Action()`
* `command.Action().ShortcutAction()`

zadjii-msft commented on src/cascadia/TerminalSettingsModel/ActionMap.cpp:
```diff
@@ -0,0 +1,329 @@
+            {
+                // This command has a name, check if it's new.
+                const auto newName{ cmd.Name() };
+                if (newName != oldCmd.Name())
```
So to be sure - this is comparing `newName` (which definitely is a given name, because `HasName` will be false for a generated name) with the current name for this command - which might be a generated name (if `oldCmd` doesn't have a name explicitly set)? 

If I do something like 

```
{ "command": { "action": "newTab" } },
{ "command": { "action": "newTab" }, "name": "foo" },
{ "command": { "action": "newTab" }, "name": "bar" },
```

will I end up with one command in the palette, or 3?

carlos-zamora replied:
One: "bar". `ActionMap` will store the command as if it was:
```js
{ "command": { "action": "newTab" }, "name": "bar" },
```

This will be _super_ useful when I start working on serialization because we won't need to write any of the older names that got overwritten.

zadjii-msft reviewed (changes requested):
Okay a bunch of stuff still needs to be done here, but I've read it all! Almost easier to just read it all like it's entirely new code, rather than with the diff. Bunch of comments below.

carlos-zamora commented on src/cascadia/TerminalSettingsModel/Command.cpp:
```diff
@@ -69,6 +75,151 @@ namespace winrt::Microsoft::Terminal::Settings::Model::implementation
+
+    hstring Command::Name() const noexcept
+    {
+        if (_name.has_value())
```
Yes! So this goes back to the problem of `IReference<hstring>` cannot be exposed in the IDL.

The current architecture is...
1. construct a `Command` for each of the commands in the JSON
2. register that `Command` in our `ActionMap`
   a. If it exists, update the existing command
   b. If it's new, just add it in like normal

The problem is, not all commands in the JSON have a `"name"`. And if the `"name"` is missing, we want to interpret that as "just use the name we were already using". Like so...
```js
// Scenario 1
// "foo" invokes the "copy" command across 3 key bindings ctrl+a/b/c
{ "name": "foo", "command": "copy", "keys": "ctrl+a" },
{                "command": "copy", "keys": "ctrl+b" },
{                "command": "copy", "keys": "ctrl+c" },

// Scenario 2
// the "paste" command has 3 key bindings ctrl+x/y/z
// the "paste" command is _NOT_ in the command palette
{ "name": "bar", "command": "paste", "keys": "ctrl+x" },
{                "command": "paste", "keys": "ctrl+y" },
{ "name": null,  "command": "paste", "keys": "ctrl+z" },
```

Storing `_name` as an `optional<string>` allows us to distinguish the two scenarios above.

TL;DR: we need to distinguish between an "inherited" name and no name again

carlos-zamora replied:
The same thing can be said for `"icon"`.

carlos-zamora commented on src/cascadia/TerminalSettingsModel/Command.cpp:
```diff
@@ -69,6 +75,151 @@ namespace winrt::Microsoft::Terminal::Settings::Model::implementation
+                    _keyMappings.erase(pos);
+                    _keyMappings.push_back(*pos);
+                    // TODO CARLOS: tests fail because this is on the wrong thread
+                    //_PropertyChangedHandlers(*this, Windows::UI::Xaml::Data::PropertyChangedEventArgs{ L"Keys" });
```
I'm just removing all of this property changed stuff for now. We're not using them since we can't modify `Command::Keys` (we just completely reconstruct a `CascadiaSettings`).

This _will_ be a problem when the Actions Page can edit `Command`s, but that's a problem for future Carlos haha.

carlos-zamora commented on src/cascadia/TerminalSettingsModel/Command.h:
```diff
@@ -51,19 +52,32 @@ namespace winrt::Microsoft::Terminal::Settings::Model::implementation
         Windows::Foundation::Collections::IMap<winrt::hstring, Model::Command> _subcommands{ nullptr };
+        std::vector<Control::KeyChord> _keyMappings;
+        std::optional<std::wstring> _name;
+        std::optional<std::wstring> _iconPath;
```
Consolidating this thread with the one above: https://github.com/microsoft/terminal/pull/9621#discussion_r612776504

carlos-zamora commented on src/cascadia/TerminalSettingsModel/ActionMap.cpp:
```diff
@@ -0,0 +1,329 @@
+            parent->_PopulateNameMap(nameMap, visitedActionIDs);
+        }
+
+        // Add NestedCommands to NameMap _after_ we handle our parents.
+        // This allows us to override whatever our parents tell us.
+        for (const auto& [name, cmd] : _NestedCommands)
+        {
+            nameMap.insert_or_assign(name, cmd);
+        }
```
Yeah, this one's trippy. Nested commands are special in that we have no way to check for equality. Thus, we can't add them to `visitedActionIDs`. Instead, we do the following:
1. In `AddAction`, remove/update `_NestedCommands` if there is a collision (i.e. `copy` action was explicitly set to have the same name as a nested command)
2. In `_PopulateNameMap`...
   a. add nested commands _last_ so that we override any collisions from our parents

Using this method, we make sure that we _do not_ modify our parents. This lets you see the state of our `ActionMap` after loading a specific settings file as opposed to the whole thing.

carlos-zamora commented on src/cascadia/TerminalSettingsModel/ActionMap.cpp:
```diff
@@ -0,0 +1,329 @@
+        }
+
+        // Update NameMap and visitedActionIDs with our parents
+        for (const auto& parent : _parents)
+        {
+            parent->_PopulateNameMap(nameMap, visitedActionIDs);
+        }
```
<details>
<summary> Train of thought</summary>

Huh, I guess I could rewrite it to be like this:
```c++
// handle parents
...

// handle my layer
for ([actionID, cmd] : _ActionMap)
{
	auto name{cmd.Name()};
	if (name)
	{
		// register name/cmd
		nameMap.insert_or_assign(name, cmd);
	}
	else
	{
		// remove the name
		nameMap.remove(name);
	}
}

// handle nested commands
...
```

Wait no... we need the action IDs to be able to remove the cmd from the name map. For example...
```js

{ "name": "foo", "command": "copy" }, // layer - parent
{ "name": "bar", "command": "copy" }, // layer - current

// Result:
// "foo" and "bar" both invoke "copy"
//
// Desired:
// "bar" --> "copy". "foo" does not exist.
```

So the "remove the name" code above would have to...
1. get the cmd
2. get the action ID from the cmd
3. iterate through the `_ActionMap` to find the old name for the matching actionID
4. remove that cmd from NameMap

`visitedActionIDs` let's us...
1. replace step 3 with a O(1) getter
2. skip over some content from our parents that isn't relevant any more
</details>

Had to sit down and think about this one haha. The main concern here is how we can prevent multiple names mapping to the same command. From the thought exercise above, it's _possible_ to do this without `visitedActionIDs`, but the benefit is that that's a quick-n-easy way to identify which action IDs have already been added and _cannot_ be removed.

carlos-zamora commented on src/cascadia/TerminalSettingsModel/ActionAndArgs.cpp:
```diff
@@ -269,7 +269,10 @@ namespace winrt::Microsoft::Terminal::Settings::Model::implementation
         }
         else
         {
-            return nullptr;
+            // Something like
+            //      { name: "foo", action: "unbound" }
+            // will _remove_ the "foo" command, by returning an "invalid" action here.
+            return make_self<ActionAndArgs>();
         }
```
Good catch!

carlos-zamora commented on src/cascadia/TerminalSettingsModel/ActionMap.cpp:
```diff
@@ -0,0 +1,329 @@
+        }
+
+        // Check our parents
+        for (const auto& parent : _parents)
```
I don't have that restriction anywhere for ActionMap. We don't really expect to ever have multiple parents here; this should scale though.

The main concern for Globals was that we were `std::move`-ing the color schemes and the actions. Once we apply some kind of layering/inheritance to color schemes, Globals should be able to scale with multiple parents, I think.

carlos-zamora commented:
So close! I just have _ONE_ failing test: `DeserializationTests::TestRebindNestedCommand`.

### Summary

#### Test
1. ✅ bind nested commands to "parent"
2. ❌ rebind "parent" to a plain "newTab" command

#### Discussion

This _should_ work. `ActionMap` is doing the following:
1. bind nested commands to "parent"
	1. deserialize the command
	1. it's a nested command!
	1. add it to the action map
		1. check if it's a nested command (it is!)
		1. add it to `_NestedCommands`
	1. when name map is requested...
		1. go through non-nested commands (none)
		1. go through nested commands...	
			1. "parent" --> <nested commands> is added
2. rebind "parent" to a plain "newTab" command
	1. deserialize the command
	1. it's a "newTab" command!
	1. add it to the action map
		1. check if it's a nested command (it isn't!)
		2. add it to the `_ActionMap`
	1. when name map is requested...
		1. go through non-nested commands (none)
		2. go through parents (**I SEE THAT THIS IS THE PROBLEM NOW**)
		3. add nested commands

Ok, I see what the problem is now. Just had to think out loud for a bit. Looks like the order is wrong in `_PopulateNameMap`. I'll marinate on a fix and submit that tomorrow morning. But other than that (and a code format), we're good to go!

carlos-zamora commented:
> So close! I just have _ONE_ failing test: `DeserializationTests::TestRebindNestedCommand`.

Fixed! I'm a bit annoyed by the fix but it works and it's not bad, tbh. Basically, I split up `_PopulateNameMap` into two functions:
1. `_PopulateNameMapWithNestedCommands`:
	- top-down approach. We're just adding the nested commands starting from the root and moving down until we get to our current layer.
2. `_PopulateNameMapWithStandardCommands`:
	- bottom-up approach. Start from the current layer, update a set of visited action IDs, then visit our parents and add any actions that we're missing.
	- This is so that we don't have to do a dance of figuring out the old name of a command when we have to update it (because we'd have to remove the old name, and add the new name).

An important assumption is that in the current layer, there is no overlap in names between the names of `_NestedCommands` and standard commands. We enforce this assumption via `ActionMap::AddAction`.

By taking a two-step approach to populating the name map, we are able to handle collisions across multiple layers.

carlos-zamora commented:
New problem: new tab dropdown doesn't display key chords.

Seems to be a problem where `GetKeyBindingForAction` isn't able to find the matching action. We're just using `HashProperty` here, so I'm gonna have to write up some tests for the hashing to make sure we're handling all of that correctly.

carlos-zamora commented:
> New problem: new tab dropdown doesn't display key chords.

Fixed! Had to move `HashProperty` into its own file `HashUtils.h` and make them all inline (Leonard suggested constexpr as a way to make them inline explicitly). This gave me more space to create specializations for `IReference`, `Color`, and other types. Doing so allows me to make sure that we always get a consistent hash from two structs composed of the same data.

While fixing this, I ran into a problem with a deserialization test. This PR introduces new behavior for actions, but I think this is the correct behavior. It is as follows:

```js
{ "name": "foo", "command": "copy" },
{ "name": "bar", "command": "copy" }
```

| | Before PR | After PR |
|--|--|--|
| NameMap keys | `foo`, `bar` | `bar` |

I believe this is the desired behavior because every command should _only_ have one name.

Using this behavior, we can fix #7441. In fact, #7441 _needs_ this. Consider the following:
```js
{ "name": "foo", "command": "copy" },
{ "name": "bar", "command": "copy" },
{ "name": null, "command": "copy" },
```

| | Before PR | After PR |
|--|--|--|
| NameMap keys | `foo`, `bar` | \<_none_\> |

The value of a null name comes from being able to find the names that map to the command, and remove them from the name map. Thus, we can leverage internal action IDs (aka hashing) to accomplish this.

zadjii-msft commented on src/cascadia/TerminalSettingsModel/ActionMap.cpp:
```diff
@@ -55,28 +55,60 @@ namespace winrt::Microsoft::Terminal::Settings::Model::implementation
+                // add a valid cmd
+                nameMap.insert_or_assign(name, cmd);
+            }
+            else
+            {
+                // remove the invalid cmd
+                nameMap.erase(name);
+            }
```
This theoretically can't get hit, right? Like, a command should never be placed into `_NestedCommands` without nested commands in the first place, right?

carlos-zamora replied:
Actually, it can! This is that special differentiation between `HasNestedCommands` and `IsNestedCommand`. In `AddAction`, we populate `_NestedCommands` with all of the `IsNestedCommand == true` commands (allowing unbinding). Here, we use `HasNestedCommands` to differentiate between those meant to unbind and actual nested commands.

zadjii-msft commented on src/cascadia/TerminalSettingsModel/Command.cpp:
```diff
@@ -46,10 +46,11 @@ namespace winrt::Microsoft::Terminal::Settings::Model::implementation
         // TODO GH#6900: We probably want ActionAndArgs::Copy here
         //              This is fine for now because SUI can't actually
         //              modify the copy yet.
```
this comment is now sus

carlos-zamora replied:
Actually, it's still valid! When we create SUI, we're still passing the original `ActionAndArgs` over 😳. But this is ok to do for now because SUI has no way of modifying actions yet. And modifying the settings.json completely reloads SUI and TermApp, so we can just push this off until `Actions.xaml` isn't a read-only page.

carlos-zamora replied:
Derp. Nevermind. Fixed.

zadjii-msft commented on src/cascadia/TerminalSettingsModel/Command.cpp:
```diff
@@ -76,6 +77,11 @@ namespace winrt::Microsoft::Terminal::Settings::Model::implementation
         return _subcommands ? _subcommands.Size() > 0 : false;
     }
 
+    bool Command::IsNestedCommand() const
```
Wait I maybe missed the thread on why we need `IsNestedCommand` separate from `HasNestedCommands`. Looks like those are only different in the case where there's a command with nested commands that failed to parse. 

Is this because we have to get everything that _might_ have nested commands first, then parse the commands, then actually check if they successfully had nested commands?

carlos-zamora replied:
Yeah, I hate everything about this. We specifically need `IsNestedCommand` for the following case:
```js
{ "name": "foo", "commands": null }
```
In this very special case, `HasNestedCommands() == false` but `IsNestedCommand() == true`.

I'll add a function comment to clarify.

zadjii-msft commented on src/cascadia/TerminalSettingsModel/Command.cpp:
```diff
@@ -303,7 +306,8 @@ namespace winrt::Microsoft::Terminal::Settings::Model::implementation
             // create an "invalid" ActionAndArgs
-            result->_action = make<ActionAndArgs>();
+            result->_ActionAndArgs = make<implementation::ActionAndArgs>();
+            result->_nestedCommand = true;
```
Oh this one right here is the reason. Maybe in the doc comment for `IsNestedCommand` we should call this out.

zadjii-msft reviewed (approved):
😌 who had the harder PR, this or #9820 😆? 

I'm done reading this, I don't hate anything in here, it fixes bugs and opens up new features. I'm down.

carlos-zamora commented on src/cascadia/TerminalSettingsEditor/Converters.idl:
```diff
@@ -59,14 +59,8 @@ namespace Microsoft.Terminal.Settings.Editor
     };
 
-    runtimeclass KeyChordToVisibilityConverter : [default] Windows.UI.Xaml.Data.IValueConverter
+    runtimeclass NullCheckToVisibilityConverter : [default] Windows.UI.Xaml.Data.IValueConverter
```
It's _techically_ used _once_ in `Actions.xaml`. If there is no key chord, we hide the key chord text block... but wait... `Actions.xaml` is already limited to actions that _have_ a key chord. So we may actually be able to remove this converter and check entirely.

carlos-zamora commented:
Bug: Actions page is missing default actions

DHowett commented:
* I set `ctrl+shift+w` to unbound but it still shows up as the official closePane shortcut in the command palette.

carlos-zamora commented:
ctrl+shift+w bug is now fixed.

Merged in and fixed the changes requiring global hotkeys.

Dismissing @zadjii-msft's review since there's been a good amount of changes (mostly due to `consolidated actions`).

lhecker commented on src/cascadia/TerminalSettingsModel/ActionMap.cpp:
```diff
@@ -0,0 +1,553 @@
+        auto actionMap{ make_self<ActionMap>() };
+
+        // copy _KeyMap (KeyChord --> ID)
+        std::for_each(_KeyMap.begin(), _KeyMap.end(), [actionMap](const auto& pair) {
+            actionMap->_KeyMap.insert(pair);
+        });
```
I'd probably make this a for loop like the other ones in this method.

lhecker commented on src/cascadia/TerminalSettingsModel/ActionMap.cpp:
```diff
@@ -0,0 +1,553 @@
+        if (!_GlobalHotkeysCache)
+        {
+            const ActionHash actionHasher{};
+            std::set<InternalActionID> visitedActionIDs{};
```
`std::unordered_set` would fit better.

lhecker commented on src/cascadia/TerminalSettingsModel/ActionMap.cpp:
```diff
@@ -0,0 +1,553 @@
+    void ActionMap::_PopulateNameMapWithStandardCommands(std::unordered_map<hstring, Model::Command>& nameMap) const
+    {
+        const ActionHash actionHasher{};
+        std::set<InternalActionID> visitedActionIDs{ actionHasher(make<implementation::ActionAndArgs>()) };
```
`std::unordered_set` would fit better.

lhecker commented on src/cascadia/TerminalSettingsModel/ActionMap.cpp:
```diff
@@ -0,0 +1,553 @@
+        FAIL_FAST_IF(_parents.size() > 1);
+        for (const auto& parent : _parents)
+        {
+            actionMap->_parents.push_back(std::move(parent->Copy()));
```
First of all: The `std::move` should be removed. Return values in C++ are special and fall into the so called "prvalue" value category. The short story is: They behave identical to a "currently being moved" value (`T&&`) and calling `std::move` on them again is actually detrimental, as this causes the variable to potentially first being put on the stack temporarily and then being moved again into the call to `push_back` (...because `std::move` needs an address and a "prvalue" doesn't have one - you can't write `auto x = &foo();`).

BTW: The only reason C++ still has `push_back()` is because they didn't want to break the stdlib ABI. That's why they called it `emplace_back()`, but technically it's `push_back2()`, or as we at MS like to say: `PushBackEx()`.
Same for `lock_guard` vs. `scoped_lock` where the former is older and takes a single parameter, while the latter and newer version is a variadic and only exists separately to not break the ABI.
`push_back` is fine as it is in this case though. In most stdlibs, including MSVC's, it's an alias for `emplace_back` only anyways.
Same is true for other stdlib containers by the way: `insert()` is implemented in terms of `emplace()`.
Don't ask me why it's `insert_or_assign` and not `emplace_or_assign`. The C++ committee often isn't quite rational to put it mildly. 😐

carlos-zamora replied:
Thanks! Went through and updated the `.insert` calls in `ActionMap` so that I can be _cool_ and _hip_ 😎

lhecker commented on src/cascadia/TerminalSettingsModel/KeyChordSerialization.cpp:
```diff
@@ -292,5 +296,57 @@ winrt::hstring KeyChordSerialization::ToString(const KeyChord& chord)
+            // "keys": "ctrl+c"
+            keyChordText = json.asString();
+        }
+        else if (json.isArray() && json.size() == 1)
```
Should there be another `json[0].isString()` check here?

lhecker commented on src/cascadia/TerminalSettingsModel/Command.cpp:
```diff
@@ -65,10 +68,158 @@ namespace winrt::Microsoft::Terminal::Settings::Model::implementation
+                else
+                {
+                    // Move the new KeyChord to the back of the line.
+                    _keyMappings.erase(pos);
+                    _keyMappings.push_back(*pos);
```
This code is incorrect and will crash, because `erase()` invalidates all iterators in the range `[pos, end())`.
--> You cannot copy `*pos`.

Fix:
```cpp
_keyMappings.erase(pos);
break;
```

Oh and: Since the prior if condition has a `return;` you can remove the `else` and unnest this by 1 (if you want to).

---

But the best improvement would be to just write it like that:
```cpp
void Command::RegisterKey(const Control::KeyChord& keys)
{
    if (!keys)
    {
        return;
    }

    EraseKey(keys);
	_keyMappings.emplace_back(keys);
}
```

lhecker commented on src/cascadia/TerminalSettingsModel/Command.cpp:
```diff
@@ -65,10 +68,158 @@ namespace winrt::Microsoft::Terminal::Settings::Model::implementation
+            if (keys.Modifiers() == pos->Modifiers() && keys.Vkey() == pos->Vkey())
+            {
+                // KeyChord was already registered.
+                if (*pos == _keyMappings.back())
```
```suggestion
                if (pos == _keyMappings.end())
```

lhecker replied:
Whoops that's wrong. This condition can never be true.

carlos-zamora commented:
This PR will not merge until #9428 merges

zadjii-msft reviewed (approved):
Yea, this looks good. ConsolidatedActions is a little hard to follow, but it makes more sense when you don't view it as a diff

DHowett reviewed (changes requested):
blocking for a cursory look

DHowett commented on src/cascadia/TerminalSettingsEditor/Actions.cpp:
```diff
@@ -19,24 +19,27 @@ namespace winrt::Microsoft::Terminal::Settings::Editor::implementation
         InitializeComponent();
 
-        _filteredActions = winrt::single_threaded_observable_vector<winrt::Microsoft::Terminal::Settings::Model::Command>();
+        _filteredActions = winrt::single_threaded_observable_vector<Command>();
```
this will become immediately obnoxious if you ever introduce `Editor::Command`

carlos-zamora replied:
I'm gonna not fix this if that's ok. I already have an Actions Page PR out. And I don't think I have that problem there because I introduced `KeyBindingViewModel` instead.

DHowett commented on src/cascadia/TerminalSettingsEditor/KeyChordConverter.h:
```diff
@@ -0,0 +1,9 @@
+#include "NullCheckToVisibilityConverter.g.h"
+#include "../inc/cppwinrt_utils.h"
+
+DECLARE_CONVERTER(winrt::Microsoft::Terminal::Settings::Editor, NullCheckToVisibilityConverter);
```
is there any way for us to simply stop introducing new converters and just like... use functions?

DHowett replied:
wait - no xaml files exist in this pull request that use this converter.

carlos-zamora replied:
Whoops! Missed that file from an earlier iteration, it seems.

DHowett commented on src/cascadia/TerminalSettingsModel/ActionAndArgs.cpp:
```diff
@@ -261,22 +261,14 @@ namespace winrt::Microsoft::Terminal::Settings::Model::implementation
+        // Something like
+        //      { name: "foo", action: "unbound" }
+        // will _remove_ the "foo" command, by returning an "invalid" action here.
+        return make_self<ActionAndArgs>(action, args);
```
I don't know how much I love this change...

carlos-zamora replied:
So, here's my reasoning, `ActionAndArgs::FromJson` _could_ use `nullptr` and `Invalid` interchangeably, yes. But then we _always_ have to check for nullness, and we'd be a bit redundant.

This cleans up the code a bit for any consumer _and_ for `ActionAndArgs::FromJson`. 😎

DHowett commented on src/cascadia/TerminalSettingsModel/ActionMapSerialization.cpp:
```diff
@@ -0,0 +1,80 @@
+    }
+
+    // Method Description:
+    // - Takes the KeyModifier flags from Terminal and maps them to the WinRT types which are used by XAML
```
nit: there's no XAML in the settings model world -- perhaps this needs a rephrase

DHowett commented on src/cascadia/TerminalSettingsModel/Command.cpp:
```diff
@@ -65,10 +68,140 @@ namespace winrt::Microsoft::Terminal::Settings::Model::implementation
 
+    // Function Description:
+    // - reports if the current command has nested commands
+    // - This CANNOT detect { "name": "foo", "commands": null }
```
and why would it? "commands":null _is the very definition of not having nested commands..._

DHowett commented on src/cascadia/TerminalSettingsModel/Command.cpp:
```diff
@@ -65,10 +68,140 @@ namespace winrt::Microsoft::Terminal::Settings::Model::implementation
+
+    // Function Description:
+    // - reports if the current command IS a nested command
+    // - This CAN be used to detect cases like { "name": "foo", "commands": null }
+    bool Command::IsNestedCommand() const noexcept
+    {
+        return _nestedCommand;
```
eehhhh? what?

carlos-zamora replied:
I hate every bit of this. The difference is...
- `HasNestedCommands`: do I contain any sub-commands
- `IsNestedCommand`: was I using the `"commands"` key in the JSON

If `HasNestedCommands` == `true`, then `IsNestedCommand` == `true`. But this is an IF, not an IFF.

So I'm using this as a way to detect `"commands": null` as a nested command that contains nothing.

DHowett commented on src/cascadia/TerminalSettingsModel/Command.cpp:
```diff
@@ -65,10 +68,140 @@ namespace winrt::Microsoft::Terminal::Settings::Model::implementation
+    // - <none>
+    void Command::EraseKey(const Control::KeyChord& keys)
+    {
+        for (auto pos = _keyMappings.begin(); pos < _keyMappings.end(); ++pos)
```
this seems like a manual version of `std::remove_if`

DHowett commented on src/cascadia/TerminalSettingsModel/Command.cpp:
```diff
@@ -174,59 +288,60 @@ namespace winrt::Microsoft::Terminal::Settings::Model::implementation
 
-        if (result->_Name.empty())
-        {
-            return nullptr;
```
this seems an important check

carlos-zamora replied:
This used to be important because `Command` was only used for the command palette. So if a command didn't have a name, it wasn't relevant to the command palette.

That's no longer the case. An Action Map will _always_ care about what you (a `Command`) have to say. So if you have no name, good for you. But we still care about your keys and ActionAndArgs.

DHowett commented on src/cascadia/TerminalSettingsModel/GlobalAppSettings.cpp:
```diff
@@ -131,10 +129,7 @@ winrt::com_ptr<GlobalAppSettings> GlobalAppSettings::Copy() const
-    {
-        globals->_keymap = _keymap->Copy();
-    }
+    globals->_actionMap = _actionMap->Copy();
```
note to self to make sure copy does parents right

DHowett commented on src/cascadia/TerminalSettingsModel/HashUtils.h:
```diff
@@ -0,0 +1,66 @@
+    template<typename T, typename... Args>
+    constexpr size_t HashProperty(const T& val, Args&&... more)
+    {
+        return HashProperty(val) ^ HashProperty(std::forward<Args>(more)...);
```
you are going to want a better hash combining algorithm than this. consider stealing the one from boost's `hash_combine`. This is a dangerous game :)

DHowett replied:
(we already ship boost, so there is no license change)

lhecker replied:
The MSVC STL ships with FNV1a which is pretty alright for such a simplistic XOR combination. So at least for hash quality boost's `hash_combine` wouldn't be an improvement per se.
IT would break the symmetry of this hash function though. At the moment `HashProperty("foo", "bar")` and `HashProperty("bar", "foo")` return the same value which indeed might be an issue.

DHowett commented on src/cascadia/TerminalSettingsModel/KeyChordSerialization.cpp:
```diff
@@ -104,14 +105,12 @@ static const std::unordered_map<std::wstring_view, int32_t> vkeyNamePairs {
 // Return Value:
 // - a newly constructed KeyChord
-KeyChord KeyChordSerialization::FromString(const winrt::hstring& hstr)
+static KeyChord _FromString(const std::wstring_view& wstr)
```
unless there's a macro forcing you into it, avoid using underscore-uppercase things

DHowett commented on src/cascadia/TerminalSettingsModel/KeyChordSerialization.cpp:
```diff
@@ -104,14 +105,12 @@ static const std::unordered_map<std::wstring_view, int32_t> vkeyNamePairs {
     std::wstring temp;
     std::vector<std::wstring> parts;
-    std::wstringstream wss(wstr);
+    std::wstringstream wss(wstr.data());
```
huh, wstringstream cannot consume a view?

carlos-zamora replied:
No. 😭

lhecker replied:
Oh I totally missed that... This is a bug!
`wstring_view` doesn't need to have a terminal null character!

Please write the following instead:
```cpp
std::wstringstream wss;
wss << wstr;
```

DHowett commented on src/cascadia/TerminalSettingsModel/KeyChordSerialization.cpp:
```diff
@@ -220,8 +219,13 @@ KeyChord KeyChordSerialization::FromString(const winrt::hstring& hstr)
 // Return Value:
 // - a string which is an equivalent serialization of this object.
-winrt::hstring KeyChordSerialization::ToString(const KeyChord& chord)
+static std::wstring _ToString(const KeyChord& chord)
```
ut supra

DHowett reviewed (changes requested):
53/59 with 14 pending -- gonna finish after this meeting

lhecker commented on src/cascadia/TerminalSettingsModel/Command.cpp:
```diff
@@ -35,16 +36,18 @@ namespace winrt::Microsoft::Terminal::Settings::Model::implementation
-        command->_IconPath = _IconPath;
+        command->_name = _name;
+        command->_ActionAndArgs = *get_self<implementation::ActionAndArgs>(_ActionAndArgs)->Copy();
+        for (const auto& keys : _keyMappings)
+        {
+            command->_keyMappings.emplace_back(keys.Modifiers(), keys.Vkey());
+        }
```
A plain copy will do the same:
```cpp
command->_keyMappings = _keyMappings;
```

carlos-zamora commented on src/cascadia/TerminalSettingsEditor/Actions.h:
```diff
@@ -9,6 +9,14 @@
 
 namespace winrt::Microsoft::Terminal::Settings::Editor::implementation
 {
+    struct CommandComparator
+    {
+        bool operator()(const Model::Command& lhs, const Model::Command& rhs) const
+        {
+            return lhs.Name() < rhs.Name();
+        }
+    };
```
Eh, keeping the struct here for a few reasons:
- Actions page will soon be using KeyBindingViewModel
- making this an operator on `Command` didn't work (I guess CppWinRT doesn't let it apply to the projected type)

DHowett commented on src/cascadia/TerminalSettingsModel/ActionMap.h:
```diff
@@ -0,0 +1,118 @@
+
+namespace winrt::Microsoft::Terminal::Settings::Model::implementation
+{
+    typedef size_t InternalActionID;
```
```suggestion
    using InternalActionID = size_t;
```

typedef is what we said when we did not know how to speak C++

C++11 came along and allowed us something a little more . . . elegant.

DHowett commented on src/cascadia/TerminalSettingsModel/ActionMap.cpp:
```diff
@@ -0,0 +1,557 @@
+    void ActionMap::_PopulateNameMapWithStandardCommands(std::unordered_map<hstring, Model::Command>& nameMap) const
+    {
+        const ActionHash actionHasher{};
+        std::unordered_set<InternalActionID> visitedActionIDs{ actionHasher(make<implementation::ActionAndArgs>()) };
```
Sorry - you're populating the action ID set with.. the hashed contents of a completely empty sentinel ActionAndArgs? why?

DHowett replied:
(also this creates an empty object just to _ask it a math question_ and delete it)

carlos-zamora replied:
`winrt::make<ActionAndArgs>()` gives us an `ActionAndArgs` with `ShortcutAction = Invalid` and `Args = nullptr`. We're prepopulating this `set` with it so that we just skip over any "invalid" actions, since those are useless to us.

Alternatively, we could do something like this:
```cpp
       std::unordered_set<InternalActionID> visitedActionIDs;
        for (const auto& cmd : _GetCumulativeActions())
        {
			// skip over all invalid actions
			if (cmd.ActionAndArgs().Action() == ShortcutAction::Invalid)
			{
				continue;
			}

            // Only populate NameMap with actions that haven't been visited already.
            const auto actionID{ Hash(cmd.ActionAndArgs()) };
            if (visitedActionIDs.find(actionID) == visitedActionIDs.end())
            {
                const auto& name{ cmd.Name() };
                if (!name.empty())
                {
                    // Update NameMap.
                    nameMap.insert_or_assign(name, cmd);
                }

                // Record that we already handled adding this action to the NameMap.
                visitedActionIDs.emplace(actionID);
            }
        }
```

DHowett commented on src/cascadia/TerminalSettingsModel/ActionMap.cpp:
```diff
@@ -0,0 +1,557 @@
+    //             There should only ever by one of each command (identified by the actionID) in the nameMap.
+    void ActionMap::_PopulateNameMapWithStandardCommands(std::unordered_map<hstring, Model::Command>& nameMap) const
+    {
+        const ActionHash actionHasher{};
```
this is kinda the thing i was talking about with the hasher just being a static function here in this file. you're not using it to implement the std::hashlike interface of any stl container.

DHowett commented on src/cascadia/TerminalSettingsModel/ActionMap.cpp:
```diff
@@ -0,0 +1,557 @@
+            const auto name{ cmd.Name() };
+            if (!name.empty())
+            {
+                _NestedCommands.Insert(name, cmd);
```
If somebody calls AddAction after either NameMap or NestedCommands or whatever, the cache will be invalid (but you're not invalidating it)

DHowett commented on src/cascadia/TerminalSettingsModel/ActionMap.cpp:
```diff
@@ -0,0 +1,557 @@
+        }
+
+        // Now check if this action was introduced in another layer.
+        const auto& consolidatedActionPair{ _ConsolidatedActions.find(actionID) };
```
it took me 300 lines to understand that "Consolidated" was the "commands that we modified which came from a lower layer" that we talked about

carlos-zamora replied:
hahaha. I hate every little bit of it. The current name is a bit too long for my liking, but I needed a way to distinguish it from the one from `_ActionMap`. :/ _Very_ open to suggestions for a different name.

For the record, the original name was `NonSerializableAction` hahaha

DHowett commented on src/cascadia/TerminalSettingsModel/ActionMap.cpp:
```diff
@@ -0,0 +1,557 @@
+        const auto& consolidatedActionPair{ _ConsolidatedActions.find(actionID) };
+        if (consolidatedActionPair != _ConsolidatedActions.end())
+        {
+            // We're adding an action that already existed on a different layer.
```
is this more, "We're adding an action that we already have an 'overwrite' record for."

carlos-zamora replied:
🍅🍅 I'll add a big ol' comment on what these actions are in the .h

DHowett commented on src/cascadia/TerminalSettingsModel/ActionMap.cpp:
```diff
@@ -0,0 +1,557 @@
+                {
+                    // Create a copy of the conflicting action,
+                    // and erase the conflicting key chord from the copy.
+                    const auto& conflictingCmdImpl{ get_self<implementation::Command>(conflictingCmd) };
+                    const auto& conflictingCmdCopy{ conflictingCmdImpl->Copy() };
```
i think for almost all of these what you really want is `auto` (not `const auto&`)

DHowett replied:
the compiler can figure it out because `const auto&` is special, but the reader needs some help :)

DHowett commented on src/cascadia/TerminalSettingsModel/ActionMap.cpp:
```diff
@@ -0,0 +1,557 @@
+                oldCmdImpl->RegisterKey(keys);
+            }
+
+            if (consolidatedCmd)
+            {
+                // Update inner Command with new key chord
+                auto consolidatedCmdImpl{ get_self<Command>(consolidatedCmd) };
+                consolidatedCmdImpl->RegisterKey(keys);
+            }
```
okay so we get here when... let me see if i can figure this out

we are updating a command from a lower layer to have keys that they didn't otherwise have, and we *did not conflict* any of our keys or the parent's keys?

I wonder if there's a clever way to handle the "copy-up and mutate" state with a helper

DHowett replied:
for each of these "Try Update" functions I'd like you to comment what state gets us into each conditional -- so somebody else can follow it later!

carlos-zamora replied:
Added a ton of examples and moved the logic around a bit. Hope this helps make it clearer!

DHowett commented on src/cascadia/TerminalSettingsModel/ActionMap.cpp:
```diff
@@ -0,0 +1,557 @@
+    Model::Command ActionMap::GetActionByKeyChord(Control::KeyChord const& keys) const
+    {
+        // Check the current layer
+        const auto& cmd{ _GetActionByKeyChordInternal(keys) };
```
`auto`

lhecker replied:
Oh it might be interesting to mention why it needs to be `auto`!
As I mentioned a couple days ago you can't take the reference to the return value of a "prvalue", which is the value class of the return value of `call()` expressions for instance. In this language it's not possible to get a pointer to something that isn't on the stack / in memory yet, so the compiler has to put it on the stack first and then puts a pointer to that stack value into "cmd".

Oh and btw: Since it's a const value any `std::move` calls will _silently_ fail. Moving stuff requires the source to be mutable, since the source is "cleared out". `std::move` will still silently work, because a `reference&&` can decay to a `reference&` and thus create a regular copy.

It's such an innocent mistake... I'll make sure to keep an eye out for this in my reviews in the future.

DHowett commented on src/cascadia/TerminalSettingsModel/ActionMap.cpp:
```diff
@@ -0,0 +1,557 @@
+    // Return Value:
+    // - the key chord that executes the given action
+    // - nullptr if the action is not bound to a key chord
+    Control::KeyChord ActionMap::GetKeyBindingForAction(ShortcutAction const& myAction, IActionArgs const& myArgs) const
```
IDL actually hates (technically) function overloads. This is fine, but just a caveat. If you want to disambiguate for IDL's sake, you can use "GetKeyBindingForActionWithArgs"

DHowett replied:
> Similarly, for overloaded methods, for the first method in a series of overloads (in lexical order), the compiler uses the original method name for the equivalent ABI interface method. Subsequent overloads are made unique by adding to the original name the smallest unused integer numeral suffix (increasing from 2).


If you really want, you can put `[method_name("GetKeyBindingForActionWithArgs")] blahblah GetKeyBindingForAction(Action, Args);` in the **idl** file, and it will internally be named GKBFAWA but WinRT consumers will see GKBFA

DHowett reviewed (changes requested):
OKAY i reviewed ActionMap.cpp

carlos-zamora commented on src/cascadia/TerminalSettingsModel/ActionMap.h:
```diff
@@ -0,0 +1,118 @@
+{
+    typedef size_t InternalActionID;
+
+    struct KeyChordHash
```
This was stolen from `KeyMapping`. We use it when we declare a key map:
```c++
std::unordered_map<Control::KeyChord, InternalActionID, KeyChordHash, KeyChordEquality> _KeyMap;
```

I think we ran into a problem before with CppWinRT already using std::hash specializations, so idk if that would work (at least right now).

carlos-zamora commented on src/cascadia/TerminalSettingsModel/ActionMap.h:
```diff
@@ -0,0 +1,118 @@
+        }
+    };
+
+    struct KeyChordEquality
```
Same as above, this was stolen from `KeyMapping` and is used in the same declarations for key map.

This one makes more sense to just have as a global operator though, so I'll do that.

carlos-zamora replied:
Ugh, I tried just that, but then all the key bindings were broken. It seems like these `unordered_map`s _need_ `KeyChordEquality`.

DHowett replied:
the unordered map should be using std::equals or std::is_equal which will use `operator==` from the type _unless it's overridden in the template args_.

Like, this should work: (writing code in godbolt)

DHowett replied:
okay, this works. it actually requires the operator to be in the leaf namespace. i could have sworn it should work in the parent namespace.

```c++
#include <unordered_map>

namespace A::B::C {
    struct Thing {
        int a, b;
    };
    // it apparently has to go here
    inline bool operator==(const C::Thing& lhs, const C::Thing& rhs) {
        return lhs.a==rhs.a && lhs.b==rhs.b;
    }
}

namespace A::B {
    struct ThingHasher {
        size_t operator()(const C::Thing& lhs) const { return 7; }
    };
    // okay, i thought you could put it here and because A::B::C is inside A::B it would find it.
}

int fleh() {
    std::unordered_map<A::B::C::Thing, int, A::B::ThingHasher> augh;
    augh[A::B::C::Thing{1,2}] = 3;
    augh[A::B::C::Thing{3,4}] = 7;
    return augh.find(A::B::C::Thing{1,2}) == augh.end();
}
```

DHowett replied:
Ah, ADL only uses the leafmost namespace for a lot of things. So yes: if you introduce an `operator==` into the MS.Terminal.Control namespace in your header file _it will get used automatically_

carlos-zamora replied:
broken = not recognized. Like, "ctrl+shift+space" wouldn't be detected.

DHowett replied:
llllet's just leave it. we can fix it in post.

carlos-zamora commented on src/cascadia/TerminalSettingsModel/ActionMap.h:
```diff
@@ -0,0 +1,118 @@
+        }
+    };
+
+    struct ActionHash
+    {
+        InternalActionID operator()(const Model::ActionAndArgs& actionAndArgs) const
```
Fixed.

Fun fact, this can't be `HashProperty` because it would then only be able to call `constexpr` functions :(.

DHowett replied:
Isn't.. isn't that just because you made it constexpr? You can make a function _not constexpr_ ;P

DHowett commented on src/cascadia/TerminalSettingsModel/ActionMap.cpp:
```diff
@@ -277,88 +302,101 @@ namespace winrt::Microsoft::Terminal::Settings::Model::implementation
                     const auto& inheritedCmdImpl{ get_self<Command>(inheritedCmd.value()) };
-                    consolidatedCmd = *inheritedCmdImpl->Copy();
-                    _ConsolidatedActions.emplace(actionID, consolidatedCmd);
+                    maskedCmd = *inheritedCmdImpl->Copy();
```
riffing on the naming thing I mentioned in Teams-
this action *masks* the one below it. It is a `masking` action :smile:

The `masked` one is `inheritedCmd` (!)

DHowett commented on src/cascadia/TerminalSettingsModel/ActionMap.cpp:
```diff
@@ -277,88 +302,101 @@ namespace winrt::Microsoft::Terminal::Settings::Model::implementation
         }
+        else
+        {
+            // We're adding an action that already existed on a different layer.
```
```suggestion
            // This is an action that we already have a mutable "masking" record for.
```

DHowett commented on src/cascadia/TerminalSettingsModel/ActionMap.cpp:
```diff
@@ -372,16 +410,16 @@ namespace winrt::Microsoft::Terminal::Settings::Model::implementation
 
-        // The same goes for consolidated actions.
-        if (consolidatedCmd)
+        // Update maskedCmd:
+        if (maskedCmd)
```
Comment suggestion for **why** we want to do this (because we just updated oldCmd, why would we have an old and a masked? explain!)

DHowett commented on src/cascadia/TerminalSettingsModel/ActionMap.h:
```diff
@@ -28,9 +28,14 @@ namespace SettingsModelLocalTests
+//inline bool operator==(const winrt::Microsoft::Terminal::Control::KeyChord& lhs, const winrt::Microsoft::Terminal::Control::KeyChord& rhs)
+//{
+//    return lhs.Modifiers() == rhs.Modifiers() && lhs.Vkey() == rhs.Vkey();
+//}
```
commented

DHowett replied:
oh i bet winrt emits these on its own too, that's probably why it failed

DHowett commented on src/cascadia/TerminalSettingsModel/ActionMap.h:
```diff
@@ -107,9 +91,18 @@ namespace winrt::Microsoft::Terminal::Settings::Model::implementation
-        std::unordered_map<InternalActionID, Model::Command> _ConsolidatedActions;
+        // Masked Actions:
+        // These are actions that were introduced in an ancestor,
+        //   but were unbound in the current layer.
```
unbound OR EDITED :smile:

DHowett reviewed (changes requested):
okay I think i understand the bulk of it

DHowett commented on src/cascadia/TerminalSettingsModel/ActionMap.cpp:
```diff
@@ -336,43 +336,43 @@ namespace winrt::Microsoft::Terminal::Settings::Model::implementation
         else
         {
-            // We're adding an action that already exists in our layer.
+            // This is an action that we already have a mutable "masking" record for.
```
but this isn't.. a masking record. is it? this is just one that *originated* in our layer. This comment was meant for line R374 ;P

DHowett reviewed (approved):
I'm comfortable with this. Thanks so much for all the comments!
